client/pdf_generator.py

```diff
@@ -0,0 +1,175 @@
+				self.logger.error('logo image specified but cannot be opened')
+				return
+
+	def make_preview(self, _):
+		outfile = self.expand_path(self.config['output_pdf'])
+		pdf_file = SimpleDocTemplate(outfile, pagesize=letter,
+	                        rightMargin=72, leftMargin=72,
+	                        topMargin=72, bottomMargin=18)
+		url = self.application.config['mailer.webserver_url']
+		pdf = self.get_template(url)
+		pdf_file.multiBuild(pdf)
+		self.logger.info('pdf preview created. Check ' + self.config['output_pdf'])
+
+	def signal_send_target(self, _, target):
+		outfile = self.expand_path(self.config['output_pdf'])
+		pdf_file = SimpleDocTemplate(outfile, pagesize=letter,
+	                    	rightMargin=72, leftMargin=72,
+	                        topMargin=72, bottomMargin=18)
+		url = self.application.config['mailer.webserver_url'] + '?uid=' + target.uid
+		pdf = self.get_template(url)
+		try:
+			pdf_file.multiBuild(pdf)
+			self.attach_pdf(outfile)
+			self.logger.info('pdf attachement made linking with uid: ' + target.uid)
+		except Exception as err:
+			self.logger.error('pdf could not be oppened: ', err)
+
+
+	def get_template(self, url):
+		logo = self.config['logo']
+		formatted_time = time.ctime()
+		company = self.application.config['mailer.company_name']
+		sender = self.application.config['mailer.source_email_alias']
+
+		Story = []
+		click_me = self.config['link_text']
+		link = '<font color=blue><link href="' + str(url) + '">' + click_me + '</link></font>'
+		if self.config['logo']:
+			im = Image(logo, 2*inch, 1*inch)
+			Story.append(im)
+
+		styles = getSampleStyleSheet()
+		styles.add(PS(name='Justify', alignment=TA_JUSTIFY))
+		ptext = '<font size=10>%s</font>' % formatted_time
+		Story.append(Spacer(1, 12))
+		Story.append(Paragraph(ptext, styles["Normal"]))
+		Story.append(Spacer(1, 12))
+		with open(self.config['template_file'], 'r') as t:
+			for line in t:
+				Story.append(Paragraph(line, styles["Normal"]))
+		Story.append(Spacer(1, 8))
+		Story.append(platypus.Paragraph(link, styles["Justify"]))
+		Story.append(Spacer(1, 12))
+		ptext = '<font size=10>Sincerely,</font>'
+		Story.append(Paragraph(ptext, styles["Normal"]))
+		Story.append(Spacer(1, 12))
+		ptext = '<font size=10>'+ sender + '</font>'
+		Story.append(Paragraph(ptext, styles["Normal"]))
+		Story.append(Spacer(1, 12))
+		ptext = '<font size=10>' + company + '</font>'
+		Story.append(Paragraph(ptext, styles["Normal"]))
+		return Story
+
+	def attach_pdf(self, outfile):
+		self.application.config['mailer.attachment_file'] = outfile
+
+	def signal_send_finished(self, _):
+		if not os.path.isfile(self.config['output_pdf']):
+			self.logger.error('no pdf file found at: ' + str(self.config['output_pdf']))
+			return
+		else:
+			self.logger.info('deleting pdf file: ' + str(self.config['output_pdf']))
+			try:
+				os.remove(self.config['output_pdf'])
+			except Exception as err:
+				self.logger.debug('Cannot delete created attachment: ', err)
```

Comment:
Same thing with error not being a string here. You could also improve the check of `os.path.isfile` above by checking if you have write permissions which IIRC are the permissions necessary to delete a file. Something like `os.path.isfile(...) and os.access(..., os.W_OK)` should do the trick.